Nama : Bifaqih Zulfa <br>
NIM : 12231945 <br>
Mata Kuliah : Big data <br>

# **B. Dataset I/O**

Pandas menyediakan berbagai method untuk membaca file tersebut hanya dengan dipanggil method itu, code yang lebih simple dan loading yang lebih, tentu saja output nya dapat berupa Series atau Dataframe.

Terdapat sangat banyak file yang dapat dibaca/dapat disimpan oleh Pandas, tapi ada beberapa file yang paling umum dan sering digunakan oleh praktisi data seperti berikut ini:

* **CSV** (*Comma Separated Values*), antar data dalam satu baris dipisahkan oleh comma, ",".
* **TSV** (*Tab Separated Values*), antar data dalam satu baris dipisahkan oleh "Tab".
* **Excel**
* **Google BigQuery**
* **SQL Query**
* **JSON** (*Java Script Object Notation*)




## **1. Read Dataset - CSV dan TSV**
**CSV** dan **TSV** pada hakikatnya adalah tipe data text dengan perbedaan terletak pada pemisah antar data dalam satu baris. Pada *file* CSV, antar data dalam satu baris dipisahkan oleh *comma*, ",". Namun, pada file *TSV* antar data dalam satu baris dipisahkan oleh "**Tab**".

Fungsi .***read_csv()*** digunakan untuk membaca file yang value-nya dipisahkan oleh comma (*default*), terkadang pemisah value-nya bisa di set ‘\t’ untuk file tsv (*tab separated values*).

**Notes** :

**Dataset csv** : https://github.com/rahmadsa/dataset/blob/main/sample_csv.csv

**Dataset tsv** : https://github.com/rahmadsa/dataset/blob/main/sample_tsv.tsv

**Membaca file CSV**

![Kode Membaca file CSV](https://github.com/rahmadsa/Data-Wrangling-With-Python/blob/main/properties/img/2024-03-2205-30-14.png?raw=true)

**Membaca file TSV**

![Kode Membaca file CSV](https://github.com/rahmadsa/Data-Wrangling-With-Python/blob/main/properties/img/2024-03-22-05-38-40.png?raw=true)





**Tugas praktek:**

Pada code cell editor dibawah ini dapat dilihat kode-kode yang tidak lengkap. Tugas kamu sekarang adalah mengganti tanda _ _ _ di code editor dengan yang sesuai seperti yang diberikan pada contoh di atas.

Tampilkanlah tiga data teratas dari kedua jenis file tersebut

In [ ]:
import pandas as pd
# File CSV
df_csv = pd.read_csv("https://raw.githubusercontent.com/rahmadsa/dataset/main/sample_csv.csv")
print(df_csv.head(3)) # Menampilkan 3 data teratas
# File TSV
df_tsv = pd.read_csv("https://raw.githubusercontent.com/rahmadsa/dataset/main/sample_tsv.tsv", sep='\t')
print(df_tsv.head(3)) # Menampilkan 3 data teratas

   order_id  order_date  customer_id             city     province product_id  \
0   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P0648   
1   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P3826   
2   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P1508   

     brand  quantity  item_price  
0  BRAND_C         4     1934000  
1  BRAND_V         8      604000  
2  BRAND_G        12      747000  
   order_id  order_date  customer_id      city     province product_id  \
0   1932339  2019-03-02        19055  Semarang  Jawa Tengah      P0648   
1   1932339  2019-03-02        19055  Semarang  Jawa Tengah      P3826   
2   1932339  2019-03-02        19055  Semarang  Jawa Tengah      P1508   

     brand  quantity  item_price  
0  BRAND_C         4     1934000  
1  BRAND_V         8      604000  
2  BRAND_G        12      747000  


penjelasan :
Kode ini menggunakan fungsi pd.read_csv() untuk membaca berkas data eksternal langsung dari URL GitHub dan mengubahnya menjadi bentuk objek Pandas DataFrame.<br>

Secara default, read_csv memisahkan data berdasarkan tanda koma (,). Namun, untuk berkas TSV (Tab Separated Values), kita perlu menambahkan argumen sep='\t' agar Pandas tahu bahwa pemisah antar kolomnya adalah karakter Tab.<br>

Fungsi .head(3) digunakan untuk menampilkan 3 baris data teratas dari masing-masing DataFrame guna memastikan data telah dimuat dengan benar.<br>

## **2. Read Dataset - Excel**
File Excel dengan ekstensi ***.xls** atau ***.xlsx** cukup banyak digunakan dalam menyimpan data. Pandas juga memiliki fitur untuk membaca file excel.

**Notes :**

**Dataset :** https://github.com/rahmadsa/dataset/blob/main/sample_excel.xlsx

Fungsi ***.read_excel()*** digunakan untuk membaca file excel menjadi dataframe pandas.

![Kode Read xls](https://github.com/rahmadsa/Data-Wrangling-With-Python/blob/main/properties/img/2024-03-2205-30-14.png?raw=true)




**Tugas praktek:**

Pada cell editor dibawah ini dapat dilihat kode-kode yang tidak lengkap. Lengkapilah kode tersebut dengan mengganti tanda _ _ _ di cell editor dengan yang sesuai seperti yang diberikan pada contoh di atas.

Tampilkanlah sepuluh data teratas dari file excel yang telah dibaca!

In [2]:
import pandas as pd
# File xlsx dengan data di sheet "test"
df_excel = pd.read_excel("https://raw.githubusercontent.com/rahmadsa/dataset/main/sample_excel.xlsx", sheet_name="test")
print(df_excel.head(5)) # Menampilkan 5 data teratas

   order_id  order_date  customer_id             city     province product_id  \
0   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P0648   
1   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P3826   
2   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P1508   
3   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P0520   
4   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P1513   

     brand  quantity  item_price  
0  BRAND_C         4     1934000  
1  BRAND_V         8      604000  
2  BRAND_G        12      747000  
3  BRAND_B        12      450000  
4  BRAND_G         3     1500000  


penjelasan :
Kode ini menggunakan fungsi pd.read_excel() untuk membaca berkas lembar kerja spreadsheet Excel (.xlsx) dari URL.<br>

Karena sebuah file Excel bisa memiliki banyak lembar kerja (sheet), parameter sheet_name="test" digunakan untuk memerintahkan Pandas agar hanya memuat data yang berada di dalam sheet yang bernama "test".<br>

## **3. Read Dataset - JSON**

Method ***.read_json()*** digunakan untuk membaca URL API yang formatnya JSON dan mengubahnya menjadi dataframe pandas. Method ini dapat digunakan seperti yang dicontohkan berikut ini:

Dataset JSON: https://github.com/rahmadsa/dataset/blob/main/covid2019-api.json

Fungsi .read_json() digunakan untuk membaca json API menjadi dataframe pandas:

![Kode](https://github.com/rahmadsa/Data-Wrangling-With-Python/blob/main/properties/img/2024-03-22-06-01-59.png?raw=true)

**Tugas praktek:**

Pada cell editor dibawah ini dapat dilihat kode-kode yang tidak lengkap. Tugas sekarang adalah mengganti tanda _ _ _ di cell editor dengan yang sesuai seperti yang diberikan pada contoh di atas.

Tampilkanlah sepuluh data teratas dari file JSON yang telah dibaca!

In [ ]:
import pandas as pd
# File JSON
url = "https://github.com/rahmadsa/dataset/blob/main/covid2019-api.json?raw=true"
df_json = pd.read_json(url)
print(df_json.head(10)) # Menampilkan 10 data teratas

                                                data          dt          ts
0  {'location': 'US', 'confirmed': 3363056, 'deat...  07-14-2020  1594684800
1  {'location': 'Brazil', 'confirmed': 1884967, '...  07-14-2020  1594684800
2  {'location': 'India', 'confirmed': 906752, 'de...  07-14-2020  1594684800
3  {'location': 'Russia', 'confirmed': 732547, 'd...  07-14-2020  1594684800
4  {'location': 'Peru', 'confirmed': 330123, 'dea...  07-14-2020  1594684800
5  {'location': 'Chile', 'confirmed': 317657, 'de...  07-14-2020  1594684800
6  {'location': 'Mexico', 'confirmed': 304435, 'd...  07-14-2020  1594684800
7  {'location': 'United Kingdom', 'confirmed': 29...  07-14-2020  1594684800
8  {'location': 'South Africa', 'confirmed': 2877...  07-14-2020  1594684800
9  {'location': 'Iran', 'confirmed': 259652, 'dea...  07-14-2020  1594684800


penjelasan : Kode ini memanfaatkan fungsi pd.read_json() untuk membaca struktur data berbasis teks format JSON (JavaScript Object Notation) langsung lewat jalur tautan web (URL API).<br>

Pandas secara otomatis melakukan parsing dari format key-value khas JSON ke dalam representasi tabel DataFrame. Pada hasilnya, struktur bersarang (nested object) di dalam kolom 'data' otomatis termuat sebagai dictionary di dalam baris DataFrame.<br>

## **4. Read Dataset - RDBMS (SQLite)**

Methode pembacaan RDBMS (SQLite) disajikan pada notebook berbeda

In [5]:
import pandas as pd
import sqlite3

# 1. Membuat database SQLite sementara kmd diisi dengan data sampel
conn = sqlite3.connect(':memory:')
df_sample = pd.DataFrame({'id': [1, 2, 3], 'nama': ['Adi', 'Budi', 'Siti'], 'Nilai': [90, 85, 88]})
df_sample.to_sql('mahasiswa', conn, index=False)

# Membaca data dari RDBMS (SQLite) menggunakan Pandas
query = "SELECT * FROM mahasiswa WHERE Nilai >= 87"
df_db = pd.read_sql_query(query, conn)

# Tampilkan hasil
print("Hasil pembacaan dari RDBMS (SQLite):")
print(df_db)

conn.close()

Hasil pembacaan dari RDBMS (SQLite):
   id  nama  Nilai
0   1   Adi     90
1   3  Siti     88


penjelasan : Fungsi utama yang digunakan adalah pd.read_sql_query(query, connection). Fungsi ini menerima dua parameter wajib: perintah bahasa SQL (SELECT * FROM ...) dan objek koneksi database (conn).

Output yang dihasilkan otomatis dikonversi oleh Pandas menjadi struktur DataFrame yang siap dianalisis.<br>

## **5. Head Dan Tail**
Seperti yang telah dipelajari sebelumnya bahwa ada method .head yang diterapkan pada suatu variabel bertipe pandas dataframe/series.

Method .head ditujukan untuk membatasi tampilan jumlah baris teratas dari dataset. Sementara itu, method .tail ditujukan untuk membatasi jumlah baris terbawah dari dataset.

Secara umum kedua method ini memiliki bentuk


```python
[nama_dataframe].head(n)
```
dan

```python
[nama_dataframe].tail(n)
```

dengan **n** merupakan jumlah baris yang akan ditampilkan, jika tidak disebutkan **n = 5** (sebagai nilai default dari n).

![Kode Head & Tails](https://github.com/rahmadsa/Data-Wrangling-With-Python/blob/main/properties/img/2024-03-22-06-15-15.png?raw=true)

**Tugas Praktek:**

**Notes :**

**Dataset csv** : https://github.com/rahmadsa/dataset/blob/main/sample_csv.csv

Berdasarkan *file sample_csv.csv* dengan alamat dataset diatas cetaklah 3 data teratas dan 3 data terbawah.



In [ ]:
import pandas as pd
# Baca file sample_csv.csv
df = pd.read_csv("https://github.com/rahmadsa/dataset/blob/main/sample_csv.csv?raw=true")
# Tampilkan 3 data teratas
print("Tiga data teratas:\n", df.head(3))
# Tampilkan 3 data terbawah
print("Tiga data terbawah:\n", df.tail(3))

Tiga data teratas:
    order_id  order_date  customer_id             city     province product_id  \
0   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P0648   
1   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P3826   
2   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P1508   

     brand  quantity  item_price  
0  BRAND_C         4     1934000  
1  BRAND_V         8      604000  
2  BRAND_G        12      747000  
Tiga data terbawah:
      order_id  order_date  customer_id      city          province product_id  \
98    1612390  2019-01-01        12681  Makassar  Sulawesi Selatan      P3354   
99    1612390  2019-01-01        12681  Makassar  Sulawesi Selatan      P3357   
100   1612390  2019-01-01        12681  Makassar  Sulawesi Selatan      P0422   

       brand  quantity  item_price  
98   BRAND_S        24      450000  
99   BRAND_S        24      450000  
100  BRAND_B         4     1325000  


Penjelasan : Fungsi .head(n) dan .tail(n) merupakan metode inspeksi cepat untuk melihat potongan data tanpa harus memuat seluruh isi tabel yang besar.<br>

.head(3) mengambil 3 baris data pertama dari indeks teratas (indeks 0, 1, 2), sedangkan .tail(3) mengambil 3 baris data terakhir dari indeks terbawah (indeks 98, 99, 100). Teknik ini sangat membantu dalam memvalidasi urutan waktu atau rentang baris dari awal hingga akhir dataset.<br>

### **6. Write Dataset**
Dalam bekerja sebagai data analisis setelah dilakukan data *cleaning dataset* yang sudah rapi tentunya disimpan terlebih dahulu ke dalam media penyimpanan.  

Pandas menyediakan fitur demikian secara ringkas melalui penerapan method pada dataframe/series yang ditabelkan berikut ini:

![Tabel Read Dataset](https://github.com/rahmadsa/Data-Wrangling-With-Python/blob/main/properties/img/2024-03-22-06-25-03.png?raw=true)

In [8]:

# Menyimpan dataframe ke dalam berbagai format file di lokal Google Colab

# 1. Save sebagai CSV
df_12231945.to_csv("tugas_bigdata_12231945.csv", index=False)

# 2. Save sebagai TSV
df_12231945.to_csv("tugas_bigdata_12231945.tsv", sep='\t', index=False)

# 3. Save sebagai JSON
df_12231945.to_json("tugas_bigdata_12231945.json", orient="records", indent=4)

print("Proses Write Dataset Berhasil! File CSV, TSV, dan JSON telah disimpan di folder Colab.")

Proses Write Dataset Berhasil! File CSV, TSV, dan JSON telah disimpan di folder Colab.


penjelasan : Setelah proses manipulasi atau pembersihan data selesai, Pandas menyediakan metode berbasis awalan .to_ untuk mengekspor DataFrame ke penyimpanan fisik.

to_csv() digunakan untuk menulis ke format teks CSV. Kita menambahkan argumen index=False agar nomor indeks baris tidak ikut tersimpan sebagai kolom baru di dalam file.

Untuk format TSV, kita tetap menggunakan fungsi to_csv() namun wajib menambahkan parameter pemisah sep='\t'.

to_json() mengubah baris-baris tabel menjadi format untaian teks JSON. Argumen orient="records" membuat formatnya rapi menjadi kumpulan objek (array of objects).<br>

**Tugas Praktek 4 :**

Carilah refrensi cara menggabungkan beberapa dataframe kedalam sebuah dataframe baru.

Berikut disajikan dua buah dataset :

1. **Dataset csv** : https://github.com/rahmadsa/dataset/blob/main/sample_csv.csv

2. **Dataset tsv** : https://github.com/rahmadsa/dataset/blob/main/sample_tsv.tsv


*Requirement* :

1. Buatlah Dataset baru yang merupakan gabungan 2 (dua) dataset diatas dalam sebuah dataframe baru dengan nama  df_NIM_Anda contoh *df_19979922*.
2. Tampilkan jumlah masing-masing data pada dataframe point no 1 (dataset 1 dan dataset 2)
3. Tampilkan jumlah data pada dataset data baru dari requirement nomor 1
4. Tampilkan 10 data teratas dan 10 data terbawah, dari data frame baru point no 1







In [6]:
#Jawaban
import pandas as pd

# Load dataset awal dari GitHub
url_csv = "https://raw.githubusercontent.com/rahmadsa/dataset/main/sample_csv.csv"
url_tsv = "https://raw.githubusercontent.com/rahmadsa/dataset/main/sample_tsv.tsv"

df_dataset1 = pd.read_csv(url_csv)
df_dataset2 = pd.read_csv(url_tsv, sep='\t')

df_12231945 = pd.concat([df_dataset1, df_dataset2], axis=0, ignore_index=True)

print("=== REQ 2: Jumlah Masing-masing Data Awal ===")
print("Jumlah baris Dataset 1 (CSV) :", len(df_dataset1), "baris")
print("Jumlah baris Dataset 2 (TSV) :", len(df_dataset2), "baris\n")

print("=== REQ 3: Jumlah Data pada Dataframe Baru ===")
print("Jumlah total baris df_12231945 :", len(df_12231945), "baris\n")

print("=== REQ 4: 10 Data Teratas (Head) ===")
print(df_12231945.head(10))
print("\n")

print("=== REQ 4: 10 Data Terbawah (Tail) ===")
print(df_12231945.tail(10))

=== REQ 2: Jumlah Masing-masing Data Awal ===
Jumlah baris Dataset 1 (CSV) : 101 baris
Jumlah baris Dataset 2 (TSV) : 101 baris

=== REQ 3: Jumlah Data pada Dataframe Baru ===
Jumlah total baris df_12231945 : 202 baris

=== REQ 4: 10 Data Teratas (Head) ===
   order_id  order_date  customer_id             city     province product_id  \
0   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P0648   
1   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P3826   
2   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P1508   
3   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P0520   
4   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P1513   
5   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P3911   
6   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P1780   
7   1612339  2019-01-01        18055  Jakarta Selatan  DKI Jakarta      P3132   
8   1612339  

Penjelasan : Untuk menggabungkan dua buah DataFrame yang memiliki struktur kolom yang sama, kita menggunakan fungsi pd.concat(). Parameter ignore_index=True ditambahkan agar indeks dari kedua tabel digabungkan secara berurutan kembali (mulai dari 0 hingga akhir), sehingga tidak terjadi duplikasi nomor indeks asal.

Berdasarkan hasil eksekusi, Dataset 1 memiliki 101 baris data dan Dataset 2 memiliki 101 baris data.

Setelah dilakukan penggabungan vertikal, jumlah total data baru di dalam tabel df_12231945 menjadi 202 baris data (101 + 101).

Fungsi .head(10) dan .tail(10) berhasil menampilkan 10 sampel baris awal dari data area Jakarta (Dataset 1) serta 10 sampel baris akhir dari data area luar Jakarta (Dataset 2).<br>